lib/hamt.ml

```diff
@@ -533,16 +533,18 @@ module Make (Config : CONFIG) (Key : Hashtbl.HashedType) :
         if f k v then (add k v yes, no) else (yes, add k v no))
       hamt (Empty, Empty)
 
-  let rec choose = function
-    | Empty -> raise Not_found
-    | Leaf (_, k, v) -> (k, v)
-    | HashCollision (_, li) -> List.hd li
-    | BitmapIndexedNode (_, base) -> choose base.(0)
-    | ArrayNode (_, children) ->
-        let rec loop n =
-          if children.(n) = Empty then loop (succ n) else children.(n)
-        in
-        choose (loop 0)
+  let choose =
+    let rec choose = function
+      | Empty -> raise Not_found
+      | Leaf (_, k, v) -> (k, v)
+      | HashCollision (_, li) -> List.hd li
+      | BitmapIndexedNode (_, base) -> choose base.(0)
+      | ArrayNode (_, children) -> choose (array_node children 0)
+    and array_node children n =
```

Comment:
Minor nitpick: what about renaming this `find_non_empty_child`? I had to read the functions again to remember what it did (which is not that hard either).